README.md
# label_detection
A ROS package that detects flat objects in an image using affine invariant feature matching

## Dependencies
affine_invariant_features
* https://github.com/yoshito-n-students/affine_invariant_features

object_detection_msgs
* https://github.com/yoshito-n-students/object_detection_msgs

## Subscribed Topics
image_raw (sensor_msgs/Image)

## Published Topics
labels_out (object_detection_msgs/Objects)

image_out (sensor_msgs/Image)
* original image on which labels are detected
* advertised and published when ~republish_image is true
* subtopics supported by image_transport are also published

## Parameters
~reference_directory (string, default: "reference")
* path to directory which contains reference feature files (usually <label_name>.yml or <label_name>.yml.gz)

~parameter_file (string, default: "parameter.yml")
* path to file which specifies feature detection algorithm (usually <feature_name>.yml)
* must be the same file used in generating reference feature files

~match_stripes (double, default: -1.0)
* number of parallel threads matching features
* -1.0 means as many as possible

~match_ratio (double, default: 0.05)
* highly depends on type of feature detection algorithm
* compared to \<number of matched features>/\<number of all features in reference>
* label is detected if ~match_ratio is lower

~area_ratio (double, default: 0.1)
* compared to \<area of label within image>/\<total area of image>
* label is detected if ~area_ratio is lower

~desired_encoding (string, default: "bgr8")
* desired image encoding for internal processing
* the encoding of an incomming image will be converted to ~desired_encoding
* must be the same as the encoding of reference images

~republish_image (bool, default: false)
* republish an image when labels on it are detected
* useful for smaller ~queue_size of external time synchronizer for image and label messages

~image_transport (string, default: "raw")
* transport type of the subscribed image topic

## Examples
label_detection_example
* https://github.com/yoshito-n-students/label_detection_example


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package label_detection
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2019-08-16)
------------------
* Initial release

include/label_detection/label_detector.hpp
#ifndef LABEL_DETECTION_LABEL_DETECTOR_HPP
#define LABEL_DETECTION_LABEL_DETECTOR_HPP

#include <stdexcept>
#include <string>
#include <vector>

#include <cv_bridge/cv_bridge.h>
#include <image_transport/image_transport.h>
#include <image_transport/publisher.h>
#include <image_transport/subscriber.h>
#include <image_transport/transport_hints.h>
#include <label_detection/label_detector_cv.hpp>
#include <nodelet/nodelet.h>
#include <object_detection_msgs/Objects.h>
#include <object_detection_msgs/cv_conversions.hpp>
#include <ros/node_handle.h>
#include <sensor_msgs/Image.h>

#include <opencv2/core.hpp>

#include <boost/foreach.hpp>

namespace label_detection {

class LabelDetector : public nodelet::Nodelet {
public:
  LabelDetector() {}

  virtual ~LabelDetector() {}

  virtual void onInit() {
    ros::NodeHandle &nh(getNodeHandle());
    ros::NodeHandle &pnh(getPrivateNodeHandle());
    image_transport::ImageTransport it(nh);

    // init detector
    detector_.init(pnh.param< std::string >("reference_directory", "reference"),
                   pnh.param< std::string >("parameter_file", "parameter.yml"));

    // load parameters
    desired_encoding_ = pnh.param< std::string >("desired_encoding", "bgr8");
    match_ratio_ = pnh.param("match_ratio", 0.05);
    area_ratio_ = pnh.param("area_ratio", 0.1);
    match_stripes_ = pnh.param("match_stripes", -1.);
    republish_image_ = pnh.param("republish_image", false);

    // setup result publishers
    if (republish_image_) {
      image_publisher_ = it.advertise("image_out", 1, true);
    }
    label_publisher_ = nh.advertise< object_detection_msgs::Objects >("labels_out", 1, true);

    // start label detection
    const image_transport::TransportHints default_hints;
    image_subscriber_ = it.subscribe(
        "image_raw", 1, &LabelDetector::onImageReceived, this,
        image_transport::TransportHints(default_hints.getTransport(), default_hints.getRosHints(),
                                        pnh /* try load pnh.resolveName(image_transport) */));
  }

private:
  void onImageReceived(const sensor_msgs::ImageConstPtr &image_msg) {
    namespace cb = cv_bridge;
    namespace odm = object_detection_msgs;

    try {
      // process the received message on demand
      if (image_publisher_.getNumSubscribers() == 0 && label_publisher_.getNumSubscribers() == 0) {
        return;
      }

      // received message to opencv image
      const cb::CvImageConstPtr image(cb::toCvShare(image_msg, desired_encoding_));
      if (!image) {
        NODELET_ERROR("Image conversion error");
        return;
      }
      if (image->image.empty()) {
        NODELET_ERROR("Empty image message");
        return;
      }

      // match features in the image and the references
      std::vector< std::string > names;
      std::vector< double > scores;
      std::vector< std::vector< cv::Point > > contours;
      detector_.detect(image->image, names, scores, contours, match_ratio_, area_ratio_, match_stripes_);
      if (names.empty() && contours.empty()) {
        // no labels found
        return;
      }

      // publish the original image for less queue size of subscribers
      // which synchronize processed images and labels
      if (republish_image_) {
        image_publisher_.publish(image_msg);
      }

      // publish names and contours of detected labels
      // (use a shared pointer to avoid data copy between nodelets)
      const odm::ObjectsPtr labels_msg(new odm::Objects);
      labels_msg->header = image_msg->header;
      labels_msg->names = names;
      labels_msg->probabilities = scores;
      labels_msg->contours = odm::toContoursMsg(contours);
      label_publisher_.publish(labels_msg);

    } catch (const std::exception &error) {
      // show runtime error when happened
      NODELET_ERROR_STREAM(error.what());
    }
  }

private:
  std::string desired_encoding_;
  double match_ratio_, area_ratio_, match_stripes_;
  bool republish_image_;

  image_transport::Subscriber image_subscriber_;
  image_transport::Publisher image_publisher_;
  ros::Publisher label_publisher_;

  LabelDetectorCv detector_;
};

} // namespace label_detection

#endif